### PR TITLE
Metal ifmr

### PR DESCRIPTION
**Pull request by samrose30**
What's Changed?

- Added an additional, metallicity dependent IFMR object to ifmr.py. This IFMR is drawn from Appendix C of Spera et. al. 2015. It is called IFMR_Spera15
- Minor changes to synthetic.py to allow the IFMR object to be called by name, and depending on whether or not it needs metallicity arguments, pass those along or not
- Modified the existing IFMR object (now called IFMR_Raithel18) with a distribution of NS masses instead of a single value. I got the distribution from Emily Ramey. 
- There is a great deal of flexibility to add additional IFMRs in the future simply by adding new IFMR objects to ifmr.py

Possible Problems
- The current MIST isochrones (or at least the ones downloaded on the Lu group computers) did not have the full range of log_ages neccessary. Specifically for non solar isochrones logage>10.01 are missing and logage<6.00 are missing. 
- Note that to use the IFMR object yoiu know must use the name of the specific object you want to use (ie IFMR_Spera15 or IFMR_Raithel18 instead of just IFMR)

### CODE REVIEW

File: spisea/ifmr.py [hunk 1]
```diff
@@ -1,27 +1,677 @@
 ##########################################################
 #
 #
-#This IFMR comes from Raithel et al. 2017
```
Comment (caseylam) [inline spisea/ifmr.py:4]: For all these, include the ads link to the published version
I think Kalirai is 2008

File: spisea/ifmr.py [hunk 2]
```diff
@@ -1,27 +1,677 @@
...
+        output_array[0][BH_idx] = rem_mass_array[BH_idx]
+        output_array[1][BH_idx] = codes['BH']
+
+        return output_array
+        
+
+
+class IFMR_Raithel18(IFMR):
+    """
+    This IFMR comes from Raithel et al. 2018
```
Comment (caseylam) [inline spisea/ifmr.py:665]: references here need to be updated

File: spisea/ifmr.py [hunk 3]
```diff
@@ -58,17 +708,10 @@ def BH_mass_low(self, MZAMS, f_ej):
     def NS_mass(self, MZAMS):
         """                                                                                                      
         Paper: 9 < MZAMS 120                                                                                     
-        Simplify to just return one value                                                                        
+        Drawing the mass from gaussian created using observational data
+        FIXME: Write up in sam's paper? I (sam) got the distribution from Emily Ramey
```
Comment (caseylam) [inline spisea/ifmr.py:712]: ask emily for the reference

File: spisea/tests/test_synthetic.py [hunk 4]
```diff
@@ -730,6 +731,355 @@ def match_model_masses2(isoMasses, starMasses):
     return
 
 
+<<<<<<< HEAD:popstar/tests/test_synthetic.py
```
Comment (caseylam) [inline spisea/tests/test_synthetic.py:734]: Did this merge right?

File: spisea/tests/test_synthetic.py [hunk 5]
```diff
@@ -730,6 +731,355 @@ def match_model_masses2(isoMasses, starMasses):
...
+    bad_idx = np.where(output_array[1] == -1)
+    WD_idx = np.where(output_array[1] == 101)
+    NS_idx = np.where(output_array[1] == 102)
+    BH_idx = np.where(output_array[1] == 103)
+
+    rem_mass = output_array[0]
+
+    return bad_idx[0], WD_idx[0], NS_idx[0], BH_idx[0], rem_mass
+
+def test_Raithel17_IFMR_1():
```
Comment (caseylam) [inline spisea/tests/test_synthetic.py:1029]: 18 righT?

File: spisea/atmospheres.py [hunk 6]
```diff
@@ -318,9 +318,9 @@ def get_cmfgenRot_atmosphere_closest(metallicity=0, temperature=24000, gravity=4
     """
     # Set up the proper root directory
     if rebin == True:
-        root_dir = '/g/lu/models/cdbs/grid/cmfgen_rot_rebin'
+        root_dir = '/g/lu/models/cdbs_20/grid/cmfgen_rot_rebin'
```
Comment (mwhosek) [inline spisea/atmospheres.py:321]: If we want to keep this function, we should remove these hard-coded paths and change them to call the environment variable PYSYN_CDBS. The PYSYN_CDBS value should replace "/g/lu/models/cdbs".

File: spisea/ifmr.py [hunk 7]
```diff
@@ -1,27 +1,677 @@
...
+        pass
+
+    def get_Z(self, Fe_H):
+        """
+        This function converts metallicity given as [Fe/H] into Z values.
+        It comes from Sharma et al. 2011 "Galaxia: A Code To Generate A Synthetic Survey of The Milky Way" equation (15) 
+        https://ui.adsabs.harvard.edu/abs/2011ascl.soft01007S/abstract
+
+        """
+        return 10**(Fe_H - 1.72125)
```
Comment (mwhosek) [inline spisea/ifmr.py:28]: I'm not sure how important it is, but this equation assumes Z_solar = 0.019. Different evolution models have different values for Z_solar, and many recent evolution models (MIST, Ekstrom+12, Parsec) have Z_solar = 0.014 or Z_solar = 0.015. This changes the value in the exponent from -1.72125 to - 1.85387. Is this difference tolerable for the rest of the IFMR calculation?

In the evolution models, the assumed value of Z_solar is saved under <evo_model_class>.z_solar. One could imagine calling that in this function if this is important.

File: spisea/synthetic.py [hunk 8]
```diff
@@ -928,6 +941,8 @@ def __init__(self, logAge, AKs, distance,
                  filters=['ubv,U', 'ubv,B', 'ubv,V',
                           'ubv,R', 'ubv,I']):
 
+        self.metallicity = metallicity #FIXME
```
Comment (mwhosek) [inline spisea/synthetic.py:944]: Has this FIXME been resolved?

File: spisea/tests/test_synthetic.py [hunk 9]
```diff
@@ -730,6 +731,355 @@ def match_model_masses2(isoMasses, starMasses):
     return
 
 
+<<<<<<< HEAD:popstar/tests/test_synthetic.py
+def test_phot_consistency(filt='all'):
```
Comment (mwhosek) [inline spisea/tests/test_synthetic.py:735]: FYI: the test_phot_consistency function has been depreciated and is no longer part of the SPISEA code base.

File: spisea/ifmr.py [hunk 10]
```diff
@@ -1,27 +1,675 @@
...
 #
-#This IFMR comes from Raithel et al. 2017
-#https://arxiv.org/pdf/1712.00021.pdf
+#IFMR_Raithel18 comes from Raithel et al. 2018 and has no metallicity dependence
+#https://ui.adsabs.harvard.edu/abs/2018ApJ...856...35R/abstract
 #
+#IFMR_Spera15 comes from Spera et al. 2015 appendix C and includes metallicity dependence
+#https://ui.adsabs.harvard.edu/abs/2015MNRAS.451.4086S/abstract
+#
+#Both IFMRs rely on Kalirai et al. 2007 WD IFMR on the low mass end < 9 M_sun for Raitehl18, and < 7 M_sun for Spera15
```
Comment (mwhosek) [inline spisea/ifmr.py:10]: Reference is Kalirai+08, right?

File: spisea/ifmr.py [hunk 11]
```diff
@@ -1,27 +1,675 @@
...
+    def get_Z(self, Fe_H):
+        """
+        This function converts metallicity given as [Fe/H] into Z values assuming Z_solar = 0.014.
+
+        """
+        return 10**(Fe_H - 1.85387)
+
+    def Kalirai_mass(self, MZAMS):
+        """                                                                                                      
+        From Kalirai+07 https://ui.adsabs.harvard.edu/abs/2008ApJ...676..594K/abstract
```
Comment (mwhosek) [inline spisea/ifmr.py:30]: Kalirai+08 ?

File: spisea/tests/test_synthetic.py [hunk 12]
```diff
@@ -730,154 +731,351 @@ def match_model_masses2(isoMasses, starMasses):
...
+
+    rem_mass = output_array[0]
+
+    return bad_idx[0], WD_idx[0], NS_idx[0], BH_idx[0], rem_mass
+
+def test_Raithel18_IFMR_1():
+    """
+    Check to make sure that the number of input objects matches the number of output objects
+    """
+    bad_idx, WD_idx, NS_idx, BH_idx, rem_mass = generate_Raithel17_IFMR()
```
Comment (mwhosek) [inline spisea/tests/test_synthetic.py:1032]: Shouldn't this be generate_Raithel18_IFMR?

**Pull request merged**